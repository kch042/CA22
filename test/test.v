`include "util.v"

// testbench
module test();

reg clk;
reg init;
wire [3:0] c;
wire [3:0] c1;
wire [3:0] c2;

// instantiate a dut (device under test)
util ut(
    .clk(clk),
    .init(init),
    .cnt(c)
);

// parameter ONE = 1;
// p p1(
//     .clk(clk),
//     .in(ONE),
//     .out(c1)
// );

// p p2(
//     .clk(clk),
//     .in(c1),
//     .out(c2)
// );



initial begin
    $dumpfile("test.vcd");
    $dumpvars(0, ut);

    clk = 0;
    init = 0;

    // initialize cnt in dut
    #1
    init = 1;
    #1

    init = 0;

    #100
    init=1;
    #1

    init=0;

end


always #10 begin
    clk = ~clk;

    // if (out == 4'b1111)
    //     $finish;
end


// in case bugs happen
always #400 begin
    $finish;
end


endmodule