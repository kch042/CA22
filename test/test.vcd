$date
	Thu Dec  8 18:42:23 2022
$end
$version
	Icarus Verilog
$end
$timescale
	1s
$end
$scope module test $end
$scope module ut $end
$var wire 1 ! clk $end
$var wire 1 " init $end
$var reg 4 # cnt [3:0] $end
$upscope $end
$upscope $end
$enddefinitions $end
#0
$dumpvars
bx #
0"
0!
$end
#1
b0 #
1"
#2
0"
#10
b1 #
1!
#20
0!
#30
b10 #
1!
#40
0!
#50
b11 #
1!
#60
0!
#70
b100 #
1!
#80
0!
#90
b101 #
1!
#100
0!
#102
b0 #
1"
#103
0"
#110
b1 #
1!
#120
0!
#130
b10 #
1!
#140
0!
#150
b11 #
1!
#160
0!
#170
b100 #
1!
#180
0!
#190
b101 #
1!
#200
0!
#210
b110 #
1!
#220
0!
#230
b111 #
1!
#240
0!
#250
b1000 #
1!
#260
0!
#270
b1001 #
1!
#280
0!
#290
b1010 #
1!
#300
0!
#310
b1011 #
1!
#320
0!
#330
b1100 #
1!
#340
0!
#350
b1101 #
1!
#360
0!
#370
b1110 #
1!
#380
0!
#390
b1111 #
1!
#400
0!
