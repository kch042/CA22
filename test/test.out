#! /opt/homebrew/Cellar/icarus-verilog/11.0/bin/vvp
:ivl_version "11.0 (stable)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/11.0/lib/ivl/system.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/11.0/lib/ivl/vhdl_sys.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/11.0/lib/ivl/vhdl_textio.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/11.0/lib/ivl/v2005_math.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/11.0/lib/ivl/va_math.vpi";
S_0x140e04280 .scope module, "p" "p" 2 23;
 .timescale 0 0;
    .port_info 0 /INPUT 1 "clk";
    .port_info 1 /INPUT 4 "in";
    .port_info 2 /OUTPUT 4 "out";
L_0x600003f80620 .functor BUFZ 4, v0x600002680240_0, C4<0000>, C4<0000>, C4<0000>;
o0x148030010 .functor BUFZ 1, C4<z>; HiZ drive
v0x600002680090_0 .net "clk", 0 0, o0x148030010;  0 drivers
o0x148030040 .functor BUFZ 4, C4<zzzz>; HiZ drive
v0x600002680120_0 .net "in", 3 0, o0x148030040;  0 drivers
v0x6000026801b0_0 .net "out", 3 0, L_0x600003f80620;  1 drivers
v0x600002680240_0 .var "x", 3 0;
E_0x600001a816b0 .event posedge, v0x600002680090_0;
S_0x140e043f0 .scope module, "test" "test" 3 4;
 .timescale 0 0;
v0x600002680480_0 .net "c", 3 0, v0x600002680360_0;  1 drivers
v0x600002680510_0 .var "clk", 0 0;
v0x6000026805a0_0 .var "init", 0 0;
S_0x140e04760 .scope module, "ut" "util" 3 13, 2 3 0, S_0x140e043f0;
 .timescale 0 0;
    .port_info 0 /INPUT 1 "clk";
    .port_info 1 /INPUT 1 "init";
    .port_info 2 /OUTPUT 4 "cnt";
    .port_info 3 /NODIR 0 "";
v0x6000026802d0_0 .net "clk", 0 0, v0x600002680510_0;  1 drivers
v0x600002680360_0 .var "cnt", 3 0;
v0x6000026803f0_0 .net "init", 0 0, v0x6000026805a0_0;  1 drivers
E_0x600001a81680 .event posedge, v0x6000026803f0_0, v0x6000026802d0_0;
    .scope S_0x140e04280;
T_0 ;
    %wait E_0x600001a816b0;
    %load/vec4 v0x600002680120_0;
    %assign/vec4 v0x600002680240_0, 0;
    %jmp T_0;
    .thread T_0;
    .scope S_0x140e04760;
T_1 ;
    %wait E_0x600001a81680;
    %load/vec4 v0x6000026803f0_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_1.0, 8;
    %pushi/vec4 0, 0, 4;
    %assign/vec4 v0x600002680360_0, 0;
    %jmp T_1.1;
T_1.0 ;
    ; show_stmt_assign_vector: Get l-value for compressed += operand
    %load/vec4 v0x600002680360_0;
    %pushi/vec4 1, 0, 4;
    %add;
    %store/vec4 v0x600002680360_0, 0, 4;
T_1.1 ;
    %jmp T_1;
    .thread T_1;
    .scope S_0x140e043f0;
T_2 ;
    %vpi_call 3 35 "$dumpfile", "test.vcd" {0 0 0};
    %vpi_call 3 36 "$dumpvars", 32'sb00000000000000000000000000000000, S_0x140e04760 {0 0 0};
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x600002680510_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x6000026805a0_0, 0, 1;
    %delay 1, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x6000026805a0_0, 0, 1;
    %delay 1, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x6000026805a0_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x6000026805a0_0, 0, 1;
    %delay 1, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x6000026805a0_0, 0, 1;
    %end;
    .thread T_2;
    .scope S_0x140e043f0;
T_3 ;
    %delay 10, 0;
    %load/vec4 v0x600002680510_0;
    %inv;
    %store/vec4 v0x600002680510_0, 0, 1;
    %jmp T_3;
    .thread T_3;
    .scope S_0x140e043f0;
T_4 ;
    %delay 400, 0;
    %vpi_call 3 67 "$finish" {0 0 0};
    %jmp T_4;
    .thread T_4;
# The file index is used to find the file name in the following table.
:file_names 4;
    "N/A";
    "<interactive>";
    "./util.v";
    "test.v";
